### PR TITLE
#100 moved pinnedKey to ApiEnvironmentType

### PR DESCRIPTION
**Pull request by tubbynl**
moved the pinnedKeys to ApiEnvironmentType

### CODE REVIEW

**Comment by OGKevin** [comment]
@tubbynl please rebase your commits to follow the following pattern. https://github.com/bunq/sdk_java/blob/7ed15ad7762b88f58de44ca1c83a4b1cdea6ae08/.zappr.yaml#L7

File: src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java [hunk 1]
```diff
@@ -5,18 +5,20 @@
  */
 public enum ApiEnvironmentType {
 
-  PRODUCTION("api.bunq.com", "v1"),
-  SANDBOX("public-api.sandbox.bunq.com", "v1");
+  PRODUCTION("api.bunq.com", "v1","sha256/nI/T/sDfioCBHB5mVppDPyLi2HXYanwk2arpZuHLOu0="),
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java]: missing space around `,`

File: src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java [hunk 2]
```diff
@@ -5,18 +5,20 @@
...
+  SANDBOX("public-api.sandbox.bunq.com", "v1","sha256/GhNvDokiMyXzhGft+xXWFGchUmmh8R5dQEnO4xu81NY=");
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java]: same

File: src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java [hunk 3]
```diff
@@ -30,4 +32,7 @@ public String getApiVersion() {
     return this.apiVersion;
   }
 
+  public String getPinnedKey() {
+    return pinnedKey;
+  }
 }
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java]: could you please the  missing new line at EOF here 🙊

File: src/main/java/com/bunq/sdk/http/ApiClient.java [hunk 4]
```diff
@@ -156,24 +151,15 @@ private void setProxyIfNeeded(OkHttpClient.Builder httpClientBuilder) {
...
-      case SANDBOX:
-        return certificateBuilder.add(
-            apiContext.getBaseUri(), PINNED_KEY_SANDBOX
-        ).build();
-      default:
-        throw new BunqException(
-            String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, apiContext.getEnvironmentType().toString())
-        );
+  private static CertificatePinner determineCertificateToPin(ApiEnvironmentType environmentType) {
+    if(environmentType!=null && environmentType.getPinnedKey()!=null) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/ApiClient.java]: missing space around operators.

File: src/main/java/com/bunq/sdk/http/ApiClient.java [hunk 5]
```diff
@@ -156,24 +151,15 @@ private void setProxyIfNeeded(OkHttpClient.Builder httpClientBuilder) {
...
+      return new CertificatePinner.Builder().add(
+              environmentType.getBaseUri(), environmentType.getPinnedKey()
+      ).build();
+    } else {
+      throw new BunqException(
+              String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, Objects.toString(environmentType,"<null>"))
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/ApiClient.java]: missing space after comma.

**Comment by OGKevin** [comment]
@tubbynl you can ignore the 80 line thing, the bot has been reconfigured and on your next push it should check for length of 100. Sorry for the inconvenience.

File: src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java [hunk 6]
```diff
@@ -5,18 +5,20 @@
  */
 public enum ApiEnvironmentType {
 
-  PRODUCTION("api.bunq.com", "v1"),
-  SANDBOX("public-api.sandbox.bunq.com", "v1");
+  PRODUCTION("api.bunq.com", "v1", "sha256/nI/T/sDfioCBHB5mVppDPyLi2HXYanwk2arpZuHLOu0="),
+  SANDBOX("public-api.sandbox.bunq.com", "v1", "sha256/GhNvDokiMyXzhGft+xXWFGchUmmh8R5dQEnO4xu81NY=");
```
[the remark this answers is not in the document]
Comment (OGKevin) [reply src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java]: this one can be reformatted to multi line.

File: src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java [hunk 7]
```diff
@@ -30,4 +32,7 @@ public String getApiVersion() {
     return this.apiVersion;
   }
 
-}
\ No newline at end of file
+  public String getPinnedKey() {
+    return pinnedKey;
+  }
+}
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/context/ApiEnvironmentType.java]: Missing new line at end of `class` block.

File: src/main/java/com/bunq/sdk/http/ApiClient.java [hunk 8]
```diff
@@ -156,24 +151,16 @@ private void setProxyIfNeeded(OkHttpClient.Builder httpClientBuilder) {
...
-        );
+  private static CertificatePinner determineCertificateToPin(ApiEnvironmentType environmentType) {
+    if(environmentType != null && environmentType.getPinnedKey() != null) {
+      return new CertificatePinner.Builder().add(
+              environmentType.getBaseUri(), environmentType.getPinnedKey()
+      ).build();
+    } else {
+      throw new BunqException(
+              String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, 
+				  Objects.toString(environmentType, "<null>"))
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/ApiClient.java]: This formatting looks weird. It should be on the same level as `String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, `

File: src/main/java/com/bunq/sdk/http/ApiClient.java [hunk 9]
```diff
@@ -156,24 +151,16 @@ private void setProxyIfNeeded(OkHttpClient.Builder httpClientBuilder) {
...
+      );
     }
   }
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/ApiClient.java]: Could you be kind enough to add the missing new line at end of the `class` block here as well 👼

File: src/main/java/com/bunq/sdk/http/ApiClient.java (hunk 8 above)
Comment (tubbynl) [reply src/main/java/com/bunq/sdk/http/ApiClient.java]: i then will be 109 chars :/

Comment (OGKevin) [reply src/main/java/com/bunq/sdk/http/ApiClient.java]: The same *indent* level as. My apologies.

File: src/main/java/com/bunq/sdk/http/ApiClient.java [hunk 10]
```diff
@@ -156,24 +152,15 @@ private void setProxyIfNeeded(OkHttpClient.Builder httpClientBuilder) {
...
-            String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, apiContext.getEnvironmentType().toString())
-        );
+  private static CertificatePinner determineCertificateToPin(ApiEnvironmentType environmentType) {
+    if (environmentType != null && environmentType.getPinnedKey() != null) {
+      return new CertificatePinner.Builder().add(
+              environmentType.getBaseUri(), environmentType.getPinnedKey()
+      ).build();
+    } else {
+      throw new BunqException(
+              String.format(ERROR_AMI_ENVIRONMENT_NOT_EXPECTED, Objects.toString(environmentType, "<null>"))
```
[the remark this answers is not in the document]
Comment (OGKevin) [reply src/main/java/com/bunq/sdk/http/ApiClient.java]: Yes this was a misunderstanding @tubbynl, your previous format was correct how ever it was not indented well.  


```
new BunqException(
    String.format(
        ERROR_AMI_ENVIRONMENT_NOT_EXPECTED,
        Objects.toString(environmentType, "<null>")
    )
)
```

**Comment by tubbynl** [comment]
fails at
`com.bunq.sdk.model.generated.endpoint.CardDebitTest > orderNewMaestroCardTest FAILED
    com.bunq.sdk.exception.PleaseContactBunqException at CardDebitTest.java:80`

but that probably also fails on a build on the `develop` branch as-is

**Comment by tubbynl** [comment]
ok, this is the end of the road for me.

i've spent too much time on applying code-formatting and all kind of (internal) code review standards, these "high" standard currently impede on fixing the larger anti-pattern existing in this implementation. rem my intention was to contribute some changes on the road away from using statics/singleton patterns. and an will share some content on why this is considered anti-pattern
* https://www.michaelsafyan.com/tech/design/patterns/singleton
* https://dzone.com/articles/singleton-anti-pattern

i'm currently setting up a poc on the bunq api and the only thing i can re-use from this project is the generated models and based on the code for the login / request / response validation i can rebuild this things using spring RestTemplate and RequestInterceptors 

i thank you guys for your time and for the shared code, i time permits  i will share my implementation which is bound to the spring framework.

close as-you-wish

**Comment by OGKevin** [comment]
@patrickdw1991 please 👁 the failing test is offline HSM.

File: src/main/java/com/bunq/sdk/context/ApiContext.java [hunk 11]
```diff
@@ -228,7 +225,7 @@ public boolean ensureSessionActive() {
 
   public boolean isSessionActive() {
     return sessionContext != null &&
-      getTimeToSessionExpiryInSeconds() < TIME_TO_SESSION_EXPIRY_MINIMUM_SECONDS;
+      getTimeToSessionExpiryInSeconds() > TIME_TO_SESSION_EXPIRY_MINIMUM_SECONDS;
```
Comment (patrickdw1991) [inline src/main/java/com/bunq/sdk/context/ApiContext.java]: @OGKevin Just to be sure are you sure that this won't create any issues?

File: src/test/java/com/bunq/sdk/http/PaginationScenarioTest.java [hunk 12]
```diff
@@ -42,9 +39,14 @@
   private static Gson gson = BunqGsonBuilder.buildDefault().create();
 
   private static void EnsureEnoughPayments() {
-    for (int i = NUMBER_ZERO; i < GetPaymentsMissingCount(); ++i) {
-      CreatePayment();
+    int paymentCount = GetPaymentsMissingCount();
```
Comment (patrickdw1991) [inline src/test/java/com/bunq/sdk/http/PaginationScenarioTest.java]: missingPaymentCount?

File: src/test/java/com/bunq/sdk/http/PaginationScenarioTest.java [hunk 13]
```diff
@@ -62,8 +64,12 @@ private static int GetPaymentsMissingCount() {
     return Payment.list(null, urlParams);
   }
 
-  private static void CreatePayment() {
-    Payment.create(new Amount(PAYMENT_AMOUNT_EUR, PAYMENT_CURRENCY), getPointerBravo(), PAYMENT_DESCRIPTION);
+  private static Payment createPayment() {
+    return  new Payment(
```
Comment (patrickdw1991) [inline src/test/java/com/bunq/sdk/http/PaginationScenarioTest.java]: Two spaces 😱

File: src/main/java/com/bunq/sdk/context/ApiContext.java (hunk 11 above)
Comment (OGKevin) [reply src/main/java/com/bunq/sdk/context/ApiContext.java]: according to my testing no, It was recreating a session on every call 😅 now it seems to behave as expected. Pipeline should not fail with 429 errors on each test anymore.

**Comment by OGKevin** [comment]
@patrickdw1991 pushed.

**Comment by OGKevin** [comment]
@andrederoos

**Pull request merged**